### DISCUSSION TITLE
"invalid character indexing" error using FeatureMatrix()

### DISCUSSION THREAD

**Discussion by zrcjessica** [stuart-lab/signac#626] [Q&A] [answered]
Hello,

I have a Signac object, `dat` that contains integrated scATAC-seq data from multiple samples. I am using `Fragments(dat[["peaks"]])` to get the list of Fragment objects stored in the Signac object. I also have a BED file with genomic ranges that I want to use to build a new peak by cell matrix. I have read in the BED file and made a GenomicRanges object, `peaks.gr`, from it. Here is my code for generating a new matrix:

```R
mtx <- FeatureMatrix(fragments = Fragments(dat[["peaks"]]), features = peaks.gr, sep = c(":","-"))
```
I am getting the following error:
```Error in intI(i, n = x@Dim[1], dn[[1]], give.dn = FALSE): invalid character indexing
Traceback:

1. FeatureMatrix(fragments = frag, features = peaks.gr, sep = c(":", 
 .     "-"))
2. sapply(X = obj.use, FUN = function(x) {
 .     SingleFeatureMatrix(fragment = fragments[[x]], features = features, 
 .         cells = cells, sep = sep, verbose = verbose, process_n = process_n)
 . })
3. lapply(X = X, FUN = FUN, ...)
4. FUN(X[[i]], ...)
5. SingleFeatureMatrix(fragment = fragments[[x]], features = features, 
 .     cells = cells, sep = sep, verbose = verbose, process_n = process_n)
6. featmat[feat.str, ]
7. featmat[feat.str, ]
8. callGeneric(x, i = i, , drop = TRUE)
9. eval(call, parent.frame())
10. eval(call, parent.frame())
11. x[i = i, , drop = TRUE]
12. x[i = i, , drop = TRUE]
13. subCsp_rows(x, i, drop = drop)
14. intI(i, n = x@Dim[1], dn[[1]], give.dn = FALSE)
15. stop("invalid character indexing")
```

Here's what my `peaks.gr` ranges look like:
```
GRanges object with 409926 ranges and 1 metadata column:
    seqnames          ranges strand |       score
       <Rle>       <IRanges>  <Rle> | <character>
  1        1   173194-173795      * |           *
  1        1   175061-175662      * |           *
  1        1   180344-180945      * |           *
  1        1   209442-210043      * |           *
  1        1   212152-212753      * |           *
  .      ...             ...    ... .         ...
  1        Y 1838154-1838755      * |           *
  1        Y 1850187-1850788      * |           *
  1        Y 1852279-1852880      * |           *
  1        Y 2827262-2827863      * |           *
  1        Y 3111656-3112257      * |           *
  -------
  seqinfo: 22 sequences from an unspecified genome; no seqlengths
```

The list of Fragments:
```
[[1]]
A Fragment object for 12461 cells

[[2]]
A Fragment object for 17011 cells

[[3]]
A Fragment object for 6235 cells

[[4]]
A Fragment object for 5935 cells

[[5]]
A Fragment object for 6154 cells

[[6]]
A Fragment object for 4381 cells

[[7]]
A Fragment object for 2948 cells

[[8]]
A Fragment object for 5608 cells

[[9]]
A Fragment object for 1137 cells

[[10]]
A Fragment object for 7954 cells

[[11]]
A Fragment object for 6611 cells

[[12]]
A Fragment object for 6850 cells

[[13]]
A Fragment object for 4606 cells
```

In the past, I have done this by creating a new Fragments object from the actual `fragments.tsv.gz` file for each sample in my integrated dataset and specifying which cells to keep by saving a list of cells retained after QC filtering from the integrated Signac object (`dat`). I then made a new peak by cell matrix for each sample (using `FeatureMatrix()`), e.g.:
```R
peaks.df <- read.table(peaks.bedfile)
peaks.gr <- GRanges(seqnames = peaks.df$V1,
	ranges = IRanges(start = peaks.df$V2, end = peaks.df$V3, names = peaks.df$V4),
	score = peaks.df$V5)

# generate new feature-bc mtx from peaks 
print("generating new feature barcode mtx")
mtx <- FeatureMatrix(fragments = fragments.file,
                    features = peaks.gr,
                    cells = bcs,
                    sep = c(":","-"))
```
Finally, I would combine all of the resulting matrices for each sample (a simple horizontal concatenation). That approached worked, so my guess is that the issue has to do with using the Fragments objects stored in the preprocessed Signac object rather than creating new Fragments object from the `fragments.tsv.gz` file. However, I don't know exactly what the issue is. I would appreciate it someone could chime in! 

Thanks!

**Comment by zrcjessica** [comment]
I'm not sure if this is potentially related to my problem, but I tried to step through the source code with the fragment file for one sample in my dataset and a few of my input features as an example, and I noticed that [`PartialMatrix()`](https://github.com/timoast/signac/blob/2e4e18354297adc4b069b48a7794aa1efed35016/R/utilities.R#L2102), which is called by ['SingleFeatureMatrix()'](https://github.com/timoast/signac/blob/2e4e18354297adc4b069b48a7794aa1efed35016/R/quantification.R#L349) (see the error message in my original post), calls `GetCellsInRegion()`. 

In the example in the docs for [`GetCellsInRegion()`](https://satijalab.org/signac/reference/getcellsinregion), the function call returns a named character vector `$cells` which contains cell barcodes containing reads overlapping the given region. The example gives one genomic region to the `region` input argument of the function, given as a string (`'chr-start-end'`). However, the invocation within `PartialMatrix()` gives `GetCellsInRegion()` a `GenomicRanges` object containing all of the features passed as input to `FeatureMatrix()`. When I ran `GetCellsInRegion()` with a sample of five ranges from my input features (`peaks.gr` in my original post), it actually gave me a list of named vectors, `$cells` and `$region`. These two lists are the same length and it appears each cell in the `$cells` vector contains reads overlapping the region in the corresponding position in the `$region` vector (i.e., neither of these lists necessarily contains unique elements). 

`PartialMatrix()` then calls [`elementNROWS()`](https://github.com/timoast/signac/blob/2e4e18354297adc4b069b48a7794aa1efed35016/R/utilities.R#L2114) which returns a value `nrep` which is treated as an integer. However, because the output of `GetCellsInRegion()` is a list of vectors, `elementNROWS()` returns a named integer vector (`cells` and `region`), leading to an error when [creating `feature.vec`](https://github.com/timoast/signac/blob/2e4e18354297adc4b069b48a7794aa1efed35016/R/utilities.R#L2152) later in the `PartialMatrix()` function:
```
Error in rep(x = seq_along(along.with = all.features), nrep): invalid 'times' argument
Traceback:

1. SingleFeatureMatrix(Fragments(dat[["peaks"]])[[1]], features = peaks.gr)
2. mylapply(X = feature.list, FUN = PartialMatrix, tabix = tbx, 
 .     cells = cells, sep = sep)   # at line 54-60 of file <text>
3. lapply(X, FUN, ...)
4. FUN(X[[i]], ...)
```

This seems weird to me, but I am also unsure if this is causing the problem I described in my original post because previously I had no issue with `FeatureMatrix()` when I first created a new a Fragments object from the fragments file instead of using the Fragments objects stored within my integrated Signac object as input to the function. I also tried stepping through the source code using a new Fragments object constructed from the fragments file which worked in the past, and that also led to the same error with `nrep`. I don't see how that difference could lead to different outcomes in the part of the package that I am describing here.

**Comment by zrcjessica** [comment]
Another update: it seems that these errors arise when I set `sep=c(":","-")` in `FeatureMatrix()`, regardless of whether I'm using the Fragments stored in my Signac object or new Fragment objects created from the `fragments.tsv.gz` files. Must have something to do with the way that the region string is formatted and how it is used to index the fragment files. Is this expected behavior?

**Reply by timoast** [reply]
Hi @zrcjessica, thanks for digging into this, I think it's a bug. I will take a look soon to see if I can reproduce the issue

**Comment by timoast** [comment, marked as answer]
This should be fixed now on the develop branch, let me know if you're still having any issues
